### PR TITLE
Feat/private listing sell to opensea

### PR DESCRIPTION
**Pull request by GopherJ**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/misc/marketplaces/SeaportAdapter.sol [hunk 1]
```diff
@@ -24,42 +24,41 @@ contract SeaportAdapter is IMarketplace {
...
-            CriteriaResolver[] memory resolvers,
-            ,
-            address recipient
-        ) = abi.decode(
-                params,
-                (AdvancedOrder, CriteriaResolver[], bytes32, address)
-            );
+        (AdvancedOrder[] memory advancedOrders, , ) = abi.decode(
+            params,
+            (AdvancedOrder[], CriteriaResolver[], Fulfillment[])
```
Comment (zhoujia6139) [inline contracts/misc/marketplaces/SeaportAdapter.sol:33]: For saving gas, if we don't need CriteriaResolver and Fulfillment, we don't need to decode them.

File: contracts/protocol/libraries/logic/MarketplaceLogic.sol [hunk 2]
```diff
@@ -469,6 +537,17 @@ library MarketplaceLogic {
...
         reserve.updateState(reserveCache);
 
+        bool willUpdateRateLater = vars.listingToken == vars.creditToken &&
+            vars.creditAmount != 0;
+        if (!willUpdateRateLater) {
+            reserve.updateInterestRates(
+                reserveCache,
+                vars.listingToken,
+                0,
+                vars.isListingTokenPToken ? 0 : vars.supplyAmount
```
Comment (billyz9611) [inline contracts/protocol/libraries/logic/MarketplaceLogic.sol]: updateInterestRates in _flashSupplyFor look like should use “liquidityAdded” not “liquidityTaken”

File: contracts/misc/marketplaces/SeaportAdapter.sol [hunk 3]
```diff
@@ -24,42 +24,41 @@ contract SeaportAdapter is IMarketplace {
 
     function getAskOrderInfo(bytes memory params)
```
Comment (billyz9611) [inline contracts/misc/marketplaces/SeaportAdapter.sol:25]: getAskOrderInfo and getBidOrderInfo most code logic are the same , could we extract the same part?

File: contracts/misc/marketplaces/SeaportAdapter.sol [hunk 4]
```diff
@@ -24,42 +24,41 @@ contract SeaportAdapter is IMarketplace {
...
-        // the person who listed NFT to sell
-        orderInfo.maker = advancedOrder.parameters.offerer;
-        require(
-            recipient == ADDRESSES_PROVIDER.getPool(),
-            Errors.INVALID_ORDER_TAKER
-        );
+        // the person who creates listing to sell NFT
+        orderInfo.maker = advancedOrders[0].parameters.offerer;
+        // the person who takes listing to buy NFT
+        orderInfo.taker = advancedOrders[1].parameters.offerer;
```
Comment (billyz9611) [inline contracts/misc/marketplaces/SeaportAdapter.sol:46]: should we remove the params.orderInfo.taker = msg.sender;  in executeBuyWithCredit  and executeBatchBuyWithCredit  ?

**Pull request merged**